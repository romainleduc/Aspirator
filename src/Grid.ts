import { Square } from './Square';

export class Grid {
  private margin;

  constructor(margin = 2) {
    this.margin = margin;
  }

  getMargin() {
    return this.margin;
  }

  initialize(width: number, height: number, squareSize: number) {
    const canvas = document.createElement('canvas');
    document.body.appendChild(canvas);
    canvas.id = 'canvas';
    canvas.width = width * (squareSize + this.margin);
    canvas.height = height * (squareSize + this.margin);

    const ctx = canvas.getContext('2d');

    if (ctx) {
      for (let y = 0; y <= height; y++) {
        for (let x = 0; x <= width; x++) {
          const square = new Square(squareSize, squareSize);
          square.setColor('#BBBBBB');
          square.placeToGrid(
            ctx,
            (squareSize + this.margin) * x,
            (squareSize + this.margin) * y
          );
        }
      }
    }
  }
}
