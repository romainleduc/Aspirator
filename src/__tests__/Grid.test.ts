import { Grid } from '../Grid';

beforeEach(() => {
  const canvas = document.getElementById('canvas');

  if (canvas) {
    canvas.remove();
  }
});

describe('Grid', () => {
  describe('initialize', () => {
    it('should create a canvas on initialization', () => {
      const grid = new Grid();
      grid.initialize(10, 10, 40);
      const canvas = document.getElementById('canvas');
      expect(canvas).not.toBeNull();
    });

    it('should initialize the size of the canvas relative to its content', () => {
      const grid = new Grid();
      grid.initialize(10, 30, 100);
      const canvas = <HTMLCanvasElement>document.getElementById('canvas');
      expect(canvas.width).toBe(10 * (100 + grid.getMargin()));
      expect(canvas.height).toBe(30 * (100 + grid.getMargin()));
    });

    it('should place the squares on the grid', () => {
      const grid = new Grid();
      grid.initialize(10, 10, 40);
      const canvas = <HTMLCanvasElement>document.getElementById('canvas');
      const ctx = <any>canvas.getContext('2d');
      const path = ctx?.__getPath();
      expect(path).toMatchSnapshot();
    });
  });
});
