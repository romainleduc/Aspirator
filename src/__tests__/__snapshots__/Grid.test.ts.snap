// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid initialize should place the squares on the grid 1`] = `
[
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  {
    "props": {
      "height": 40,
      "width": 40,
      "x": 420,
      "y": 420,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "rect",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "closePath",
  },
]
`;
